File: Lab01/bfs.java
import java.util.*;

public class bfs {

    private int vertices;
    private LinkedList<Integer> adjList[];

    // Constructor
    bfs(int v) {
        vertices = v;
        adjList = new LinkedList[v];

        for (int i = 0; i < v; i++) {
            adjList[i] = new LinkedList<>();
        }
    }

    // Add edge (Undirected)
    void addEdge(int v, int w) {
        adjList[v].add(w);
        adjList[w].add(v);
    }

    // BFS Method
    void BFS(int start) {
        boolean visited[] = new boolean[vertices];//array to keep track of visited vertices
        Queue<Integer> queue = new LinkedList<>();//queue to process vertices in BFS order

        visited[start] = true;
        queue.add(start);

        System.out.print("BFS Traversal: ");

        //while queue is not empty, remove the front vertex and print it
        while (!queue.isEmpty()) { 
            int v = queue.poll();
            System.out.print(v + " ");

            for (int neighbor : adjList[v]) {
                if (!visited[neighbor]) {
                    visited[neighbor] = true;
                    queue.add(neighbor);
                }
            }
        }
    }

    public static void main(String args[]) {

        bfs g = new bfs(9);

        // g.addEdge(0, 1);
        // g.addEdge(0, 2);
        // g.addEdge(1, 3);
        // g.addEdge(2, 4);
        // g.addEdge(3, 5);
        // g.addEdge(3, 6);
        // g.addEdge(4, 7);
        // g.addEdge(4, 8);

        g.addEdge(1, 3);
        g.addEdge(1, 5);
        g.addEdge(3, 7);
        g.addEdge(3, 8);
        g.addEdge(5, 6);
        g.addEdge(0, 1);
        g.addEdge(0, 2);
        g.addEdge(2, 4);        

        g.BFS(0);
    }
}

File: Lab01/bfs.py
from collections import deque


def bfs_recursive(graph, queue, visited):
    if not queue:
        return 


    vertex = queue.popleft()
    print(vertex, end=" ")

    for neighbor in graph[vertex]:
        if neighbor not in visited:
            visited.add(neighbor)
            queue.append(neighbor)

    
    bfs_recursive(graph, queue, visited)



def bfs(graph, start):
    visited = set()
    queue = deque()

    visited.add(start)
    queue.append(start)

    print("BFS Traversal:")
    bfs_recursive(graph, queue, visited)



graph = {
    'A': ['B', 'C'],
    'B': ['A', 'D', 'E'],
    'C': ['A', 'F'],
    'D': ['B'],
    'E': ['B', 'F'],
    'F': ['C', 'E']
}


bfs(graph, 'B')